server/src/test/java/org/elasticsearch/cluster/metadata/DataStreamLifecycleTests.java

```diff
@@ -135,7 +136,8 @@ public void testXContentSerializationWithRolloverAndEffectiveRetention() throws
             } else {
                 assertThat(serialized, containsString("data_retention"));
             }
-            if (lifecycle.isEnabled()) {
+            if (lifecycle.isEnabled()
+                && (systemDataStream == false || (lifecycle.getDataRetention() != null && lifecycle.getDataRetention().value() != null))) {
                 assertThat(serialized, containsString("effective_retention"));
             } else {
                 assertThat(serialized, not(containsString("effective_retention")));
```

Comment:
I'm having trouble understanding why this test hasn't been failing because I'd think it would be possible for `globalRetention` and `lifecycle.getDataRetention` to both be `null`, causing the lifecycle to not have any `effective_retention`, right? Either I'm missing something here or this just hasn't failed _yet_ (although the latter seems unlikely as this test has been succeeding for a while now).

Reply:
Ha it looks like the reason is what might be a bug here: https://github.com/elastic/elasticsearch/blob/main/server/src/test/java/org/elasticsearch/cluster/metadata/DataStreamGlobalRetentionTests.java#L81. That method will never return a retention that has both a null default and a null max.

Reply:
Ahh I was (incorrectly) looking at https://github.com/elastic/elasticsearch/blob/fec267a770470fa2756731ff2fabd61b3548bc42/test/framework/src/main/java/org/elasticsearch/cluster/metadata/DataStreamTestHelper.java#L390-L392 instead of the one you linked. That answers it indeed :)

Reply:
Good catch. I created https://github.com/elastic/elasticsearch/pull/108113 to fix that.